Add documentation about how to add new pages to your client side (#25)
front-commerce/developers.front-commerce.com: source/docs/essentials/add-a-page-client-side.md

---
id: add-a-page-client-side
title: Add a new page
---

Front-Commerce comes with a classic set of pages for an e-commerce application.
It has URLs for Catalog routes (Categories, Product, Search…), Checkout routes,
Account routes, Cms pages…

> But when you are building your own e-commerce experience you will most likely need

Reviewer: ```suggestion
But when building your own e-commerce experience you will most likely need
```

  </blockquote>
* `my-module/web/theme/pages/Recipes/index.js`: will proxy the `Recipes.js` file in order to be able to do imports on the folder directly. See [this blog post](http://bradfrost.com/blog/post/this-or-that-component-names-index-js-or-component-js/) for more context about this pattern.
<!-- TODO add comment about code splitting and link to our documentation -->
```jsx
import Recipes from './Recipes.js'

export default Recipes;
```
* `my-module/web/theme/pages/Recipes/Recipes.story.js`: will add your page to the Storybook of your application. In the case of a page it might be troublesome to setup, but it is worth it because it will improve your code quality by being in a sandboxed environment and it will help you reference the edge cases.
```jsx
import Recipes from "./Recipes.js";
import { storiesOf } from "@storybook/react";

storiesOf("pages.Recipes", module).add("default", () => {
  return <Recipes />;
});
```
  <blockquote class="note">
  For a more detailed explanation of how Storybook works in the context of Front-Commerce, please refer to <a href="#TODO">Add a component to Storybook</a>.
  </blockquote>

### Map the URL to the page component

> Once you have created your component, you need to actually tell the application how to use it. To do so, we provide an extension point that will let you add your custom routes.

Reviewer: ```suggestion
Once you have created your component, you need to actually tell the application how to use it. To do so, Front-Commerce provides an extension point that will let you declare custom routes.
```

```jsx
import React from "react";
import Route from "react-router/Route";
import Recipes from "theme/pages/Recipes";

export default () => [
  <Route
    path="/recipes"
    exact
    render={() => <Recipes />}
    key="recipes"
  />
];
```

Once you've created your file, you can refresh your application
(`npm run start`), and you should see your new route if you go
to the `/recipes` URL.

🎉

<blockquote class="warning">
    **Warning:** If several modules are registered in your application and several of them define the same route, the page of the first module in `.frontcommerce.js` will be displayed. This can be useful if you want to override a default feature of Front-Commerce.
</blockquote>

## Add a page using an URL with parameters

Now, let's add a `RecipeDetails` page that should be mapped to the
`/recipes/:recipe-id-or-slug`. You will need to follow the exact
same steps as for the `/recipes` page:
* Create a `RecipeDetails` component in the
`theme/pages/Recipes/RecipeDetails` folder
* Map the `/recipes/:recipe-id-or-slug` to the new component

> But how can you make the URL have a parameter (`:recipe-id-or-slug`)?

Reviewer: ```suggestion
**But how can you make the URL have a parameter (`:recipe-id-or-slug`)?**
```

The question is not very visible, I think it should stand out… either with bold as suggested or using a `<blockquote>`

In fact, if you take a look at the line #2 of the `moduleRoutes.js` file we have just created, you will notice that we are using the [React Router](https://github.com/ReactTraining/react-router) routing library. For this reason, you can use any feature that it provides. In the `RecipesDetails` case, you can use the params syntax of React Router as shown below.

```diff
import React from "react";
import Route from "react-router/Route";
import Recipes from "theme/pages/Recipes";
+import RecipeDetails from "theme/pages/Recipes/RecipeDetails";

export default () => [
  <Route
    path="/recipes"
    exact
    render={() => <Recipes />}
    key="recipes"
-  />
+  />,
+  <Route
+    path="/recipes/:slug"
+    render={({ match }) => <RecipeDetails slug={match.params.slug} />}
+    key="recipe-details"
+  />
];
```

You will then be able to use the `slug` as a property of your `RecipeDetails` page like any other property.

## What about dynamic URLs?

If your URL can't be mapped to a pattern that React Router can recognize, do not worry,

> we have got yourself covered! There is a second extension point in Front-Commerce that let you do this: the `Dispatcher`.

Reviewer: ```suggestion
we have got you covered! There is a second extension point in Front-Commerce that allows you to achieve this: the `Dispatcher`.
```

---
id: add-a-page-client-side
title: Add a new page
---

Depending on the level of customization of your Front-Commerce application, you might want to add a new kind of page to your React application.

There are two ways to do so:

1. add a page using a static URL
2. add a page using dynamic URLs

## Add a page using a static URL

Let's say that your goal is to display a component `MyCustomPage` when a user goes to the URL `/custom-page`.

To do so, you need to create the file `my-module/web/moduleRoutes.js` in your module. (See [Extend the theme](/docs/essentials/extend-the-theme.html) if you don't have any module yet.)

```jsx
import React from "react";
import Route from "react-router/Route";
import MyCustomPage from "theme/pages/MyCustomPage";

export default () => [
  <Route path="/custom-page" exact render={() => <MyCustomPage />} key="1" />
];
```

> If you take a look at the line #2, you will notice that the underlying technology is [React Router](https://github.com/ReactTraining/react-router). For this reason, you can use any feature that is available in this library. In particular, if you need to pass some paramaters in your URL, you can use the params the params syntax of React Router (`/page/:id`).

Reviewer: ```suggestion
If you take a look at the line #2, you will notice that we are using the [React Router](https://github.com/ReactTraining/react-router) routing library. For this reason, you can use any feature that it provides. In particular, if you need to pass some parameters in your URL, you can use the params syntax of React Router (`/page/:id`).
```

<blockquote class="warning">
    **Warning:** If several modules are registered in your application and several of them define the same route, the page of the first module in `.frontcommerce.js` will be displayed.
</blockquote>

## Add a page using dynamic URLs

Some URLs might be a bit more troublesome to use. For instance, in a e-commerce shop, you usually want some nice URLs for SEO reasons. Instead of `/product/:sku` you'll want `/my-product-title`. If you come from a Magento background, this is what we call [URL Rewrites](https://docs.magento.com/m2/ce/user_guide/marketing/url-rewrite.html).

Some of these cases are already taken into account within Front-Commerce (Products, Categories, CMS pages, ...). But depending on your own site, you might need to add new ones.

To do so, you need to:

1. Export the named object `dispatchedRoutes` in your `my-module/web/moduleRoutes.js` file

> ```

Reviewer: ```suggestion
```js
```

export const dispatchedRoutes = {

>   pageType: props => <MyCustomPage id={props.matched.identifier} />

Reviewer: We should show the whole module as above IMO (with imports) for more consistency.
